web3_proxy/src/frontend/users/payment_stripe.rs

```diff
@@ -0,0 +1,236 @@
+};
+use ethers::abi::AbiEncode;
+use ethers::types::{Address, Block, TransactionReceipt, TxHash, H256};
+use hashbrown::{HashMap, HashSet};
+use http::{HeaderMap, StatusCode};
+use migration::sea_orm::prelude::Decimal;
+use migration::sea_orm::{
+    self, ActiveModelTrait, ActiveValue, ColumnTrait, EntityTrait, IntoActiveModel, ModelTrait,
+    QueryFilter, QuerySelect, TransactionTrait,
+};
+use migration::{Expr, OnConflict};
+use payment_contracts::ierc20::IERC20;
+use payment_contracts::payment_factory::{self, PaymentFactory};
+use serde_json::json;
+use std::num::NonZeroU64;
+use std::sync::Arc;
+use stripe::Webhook;
+use tracing::{debug, error, info, trace};
+
+/// `GET /user/balance/stripe` -- Use a bearer token to get the user's balance and spend.
+///
+/// - shows a list of all stripe deposits, all fields from entity
+#[debug_handler]
+pub async fn user_stripe_deposits_get(
+    Extension(app): Extension<Arc<Web3ProxyApp>>,
+    TypedHeader(Authorization(bearer)): TypedHeader<Authorization<Bearer>>,
+) -> Web3ProxyResponse {
+    let (user, _semaphore) = app.bearer_is_authorized(bearer).await?;
+
+    let db_replica = app.db_replica().context("Getting database connection")?;
+
+    // Filter by user ...
+    let receipts = stripe_increase_balance_receipt::Entity::find()
+        .filter(increase_on_chain_balance_receipt::Column::DepositToUserId.eq(user.id))
+        .all(db_replica.as_ref())
+        .await?;
+
+    // Return the response, all except the user ...
+    let response = json!({
+        "user": Address::from_slice(&user.address),
+        "deposits": receipts,
+    });
+
+    Ok(Json(response).into_response())
+}
+
+/// `POST /user/balance/stripe/` -- Process a stripe transaction;
+/// this endpoint is called from the webhook with the user_id parameter in the request
+#[debug_handler]
+pub async fn user_balance_stripe_post(
+    Extension(app): Extension<Arc<Web3ProxyApp>>,
+    ip: InsecureClientIp,
+    headers: HeaderMap,
+    Path(mut params): Path<HashMap<String, String>>,
+    bearer: Option<TypedHeader<Authorization<Bearer>>>,
+) -> Web3ProxyResponse {
+    // rate limit by bearer token **OR** IP address
+    let (authorization, _semaphore) = if let Some(TypedHeader(Authorization(bearer))) = bearer {
+        let (_, semaphore) = app.bearer_is_authorized(bearer).await?;
+
+        // TODO: is handling this as internal fine?
+        let authorization = Web3ProxyAuthorization::internal(app.db_conn())?;
+
+        (authorization, Some(semaphore))
+    } else {
+        let InsecureClientIp(ip) = ip;
+        let authorization = login_is_authorized(&app, ip).await?;
+        (authorization, None)
+    };
+
+    let recipient_user_id: u64 = params
+        .remove("user_id")
+        .ok_or(Web3ProxyError::BadRouting)?
+        .parse()
+        .or(Err(Web3ProxyError::ParseAddressError))?;
+
+    // Get the payload, and the header
+    let payload = params.remove("data").ok_or(Web3ProxyError::BadRequest(
```

Comment:
Is the payload actually in the headers? Isn't it in the POST body?

Reply:
according to this example, data is in payload, and the signature is in the header: https://stripe.com/docs/webhooks

Reply:
Ok. This is wrong then. This needs to use the `Json` body extractor instead of the Params extractor